src/azul/indexer/__init__.py

```diff
@@ -59,6 +67,101 @@ class BundleFQID(SupportsLessThan):
     version: BundleVersion
 
 
+@attr.s(frozen=True, auto_attribs=True, kw_only=True)
+class Prefix:
+    common: str = ''
+    partition: Optional[int]
+
+    def __attrs_post_init__(self):
+        validate_uuid_prefix(self.common)
+        assert ':' not in self.common, self.common
+        if self.partition:
+            assert isinstance(self.partition, int), self.partition
+
+    @classmethod
+    def parse(cls, prefix: str) -> 'Prefix':
+        """
+        >>> Prefix.parse('aa/1')
+        Prefix(common='aa', partition=1)
+
+        >>> p = Prefix.parse('aa')
+        >>> p
+        Prefix(common='aa', partition=None)
+        >>> str(p)
+        'aa'
+        >>> from unittest.mock import patch
+        >>> import os
+        >>> with patch.dict(os.environ, AZUL_PARTITION_PREFIX_LENGTH='2'):
+        ...     p.effective
+        Prefix(common='aa', partition=2)
+        >>> Prefix.parse('aa/')
+        Traceback (most recent call last):
+        ...
+        azul.RequirementError: Prefix source: `aa/` cannot end in a `/`.
+        >>> Prefix.parse('8-/2')
+        Traceback (most recent call last):
+        ...
+        azul.RequirementError: Remove the trailing `-` from the uuid prefix: 8-
+        """
+        source_delimiter = '/'
+        reject(prefix.endswith(source_delimiter),
+               f'Prefix source: `{prefix}` cannot end in a `{source_delimiter}`.')
+        if prefix == '':
+            entry = ''
+            partition = None
+        else:
+            try:
+                entry, partition = prefix.split(source_delimiter)
+            except ValueError:
+                entry = prefix
+                partition = None
+            if partition:
+                try:
+                    partition = int(partition)
+                except ValueError:
+                    raise ValueError('Partition prefix length must be an integer: ',
+                                     partition)
+        validate_uuid_prefix(entry)
+        return cls(common=entry, partition=partition)
+
+    @cached_property
+    def effective(self) -> 'Prefix':
+        if self.partition is None:
+            return attr.evolve(self, partition=config.partition_prefix_length)
+        else:
+            return self
+
+    @classmethod
+    def of_everything(cls) -> 'Prefix':
+        return cls(common='', partition=config.partition_prefix_length)
+
+    def partition_prefixes(self) -> Iterator[str]:
+        """
+        >>> list(Prefix.parse('/0').partition_prefixes())
+        ['']
+        >>> sorted(Prefix.parse('/1').partition_prefixes())
+        ['0', '1', '2', '3', '4', '5', '6', '7', '8', '9', 'a', 'b', 'c', 'd', 'e', 'f']
+        >>> len(set(Prefix.parse('/2').partition_prefixes()))
+        256
+        >>> sorted(Prefix.parse('8f538f53/1').partition_prefixes())
+        Traceback (most recent call last):
```

Comment:
This exception is raised by `parse`, not `partition_prefixes` so this doctest case should be in `parse` not here.

Reply:
The exception is not being raised by `parse`, its being raised within `partition_prefixes`, take a look at the 
error that is raised: `azul.uuids.InvalidUUIDPrefixError: '8f538f530' is not a valid UUID prefix.` The 0 was not present within the common prefix, it was added by the map during the method, which then fails to validate before its yielded out.